cli/src/daemon.rs

```diff
@@ -268,11 +336,46 @@ impl<'a, 'b> Daemon<'a, 'b> {
     }
 
     /// Sleep either for the configured poll interval, or until it is our maintainer duty.
-    ///
-    /// TODO(ruuda): Implement the sleeping until maintainer duty part.
     fn sleep_until_next_iteration(&mut self) {
-        let sleep_time = Duration::from_secs(*self.opts.max_poll_interval_seconds());
-        println!("Sleeping {:?} until next iteration ...", sleep_time);
+        let maintainer = self.config.signer.pubkey();
+        let poll_interval = Duration::from_secs(*self.opts.max_poll_interval_seconds());
+
+        // Find out when our next maintainer duty slice starts (if any), and
+        // estimate how long it will take (after the previous snapshot publish,
+        // but this method should be called right after) until then.
+        let next_duty_slot = self
+            .snapshot_mutex
+            .lock()
+            .unwrap()
+            .as_ref()
+            .and_then(|snapshot| snapshot.solido.as_ref())
+            .and_then(|solido| solido.get_next_maintainer_duty_slot(&maintainer));
+
+        let sleep_time = next_duty_slot
+            .and_then(|slot| self.block_time_estimator.estimate_time_until_slot(slot))
+            .unwrap_or(poll_interval)
+            .min(poll_interval);
+
+        fn fmt_option<T: std::fmt::Debug>(opt_value: Option<T>) -> String {
+            opt_value
+                .map(|x| format!("{:?}", x))
+                .unwrap_or("n/a".to_string())
+        }
+
+        fn fmt_option_duration(opt_value: Option<Duration>) -> String {
+            opt_value
+                .map(|x| format!("{:.3}s", x.as_secs_f32()))
+                .unwrap_or("n/a".to_string())
+        }
+
+        println!(
+            "Sleeping until next iteration. Slot: {}, next duty slot: {}, block time: {}, sleep time: {}",
```

Comment:
One addition I would make is to prefix the log with something like "ON-DUTY" and "OFF-DUTY" so that it is easier to discern rather than trying to match the slot numbers.

Reply:
That’s a nice idea, added! It now looks like this:

```
OFF-DUTY Sleeping until next iteration. Slot: 81458907, next duty slot: 81459000, block time: n/a, sleep time: 30.000s
OFF-DUTY Sleeping until next iteration. Slot: 81458980, next duty slot: 81459000, block time: 0.414s, sleep time: 8.277s
ON-DUTY  Sleeping until next iteration. Slot: 81459000, next duty slot: 81459200, block time: 0.414s, sleep time: 30.000s
ON-DUTY  Sleeping until next iteration. Slot: 81459076, next duty slot: 81459200, block time: 0.407s, sleep time: 30.000s
OFF-DUTY Sleeping until next iteration. Slot: 81459153, next duty slot: 81459200, block time: 0.402s, sleep time: 18.908s
OFF-DUTY Sleeping until next iteration. Slot: 81459199, next duty slot: 81459200, block time: 0.404s, sleep time: 0.404s
OFF-DUTY Sleeping until next iteration. Slot: 81459199, next duty slot: 81459200, block time: 0.405s, sleep time: 0.405s
ON-DUTY  Sleeping until next iteration. Slot: 81459200, next duty slot: 81459400, block time: 0.406s, sleep time: 30.000s
ON-DUTY  Sleeping until next iteration. Slot: 81459277, next duty slot: 81459400, block time: 0.405s, sleep time: 30.000s
OFF-DUTY Sleeping until next iteration. Slot: 81459353, next duty slot: 81459400, block time: 0.403s, sleep time: 18.962s
OFF-DUTY Sleeping until next iteration. Slot: 81459399, next duty slot: 81459400, block time: 0.403s, sleep time: 0.403s
ON-DUTY  Sleeping until next iteration. Slot: 81459400, next duty slot: 81459600, block time: 0.402s, sleep time: 30.000s
ON-DUTY  Sleeping until next iteration. Slot: 81459476, next duty slot: 81459600, block time: 0.404s, sleep time: 30.000s
OFF-DUTY Sleeping until next iteration. Slot: 81459551, next duty slot: 81459600, block time: 0.406s, sleep time: 19.886s
```

Reply:
Epic!